### ISSUE TITLE
Beyond BigGAN: all-attention GANs?

### ISSUE THREAD

**Issue by gwern** [tensorfork/tensorfork#31]
Something to consider as a highly speculative research project once Ganbooru is done.

---

A distinct trend in recent DL has been self-attention moving beyond sequence or text data to image data as well, increasingly augmenting or replacing convolutions (["attention is all you need"](https://arxiv.org/pdf/1706.03762v3.pdf)). Examples:

- https://ai.facebook.com/blog/making-transformer-networks-simpler-and-more-efficient/
- https://arxiv.org/abs/1904.09925
- https://openreview.net/forum?id=HkxaFoC9KQ
- https://arxiv.org/abs/1911.03584
- https://arxiv.org/abs/1911.12287
- https://arxiv.org/abs/2001.04451
- https://arxiv.org/abs/1904.10509

In Bello et al, the classifier using solely attention is both parameter and FLOP-efficient compared to the classical CNN, and although it doesn't set the SOTA, the performance curves look good. Another paper argues that self-attention learns convolution-like attention patterns, but they are are different and more tailored; but still other papers note that NNs with attention appear to have better inductive biases, generalize in more human-like ways, and resist adversarial examples better; and of course, we've long noted in GANs that attention appears to help a lot in global coherency.

So the logical question becomes, can we create a GAN which drops convolution layers entirely in favor of just self-attention layers? The self-attention may require a lot of data in order to surpass the convolution hardwired priors & compute-efficiency but it will ultimately deliver better results. (Making it another bitter lesson #28?). Fortunately, we have both a lot of data and a lot of compute in TPU pods.

Bringing this up on Twitter, no one has a solid theoretical argument why an all-attention GAN would fail, other than the obvious point that the memory consumption of straightforward 256px or 512px self-attention layers may be too large for TPUs.

We can work around that with reversible attention layers to save memory, use of more efficient self-attention approximations like Reformer, or by rethinking the architecture entirely: why do we need 256px or 512px upscaling modules at all? You wouldn't necessarily have to do the obvious approach of using self-attention+nearest-neighbors-upscaling blocks iterated (16→32→64→128→256→512px). You could have a deep stack of small self-attention layers, and then do the pixel generation in two or three big blocks. Why not use something like the deep _w_ stack of StyleGAN, (#26) where the self-attention layers keep transforming a latent representation of the image, before spitting out the actual pixels? This would be a 'skinny' GAN. You could see this as taking the StyleGAN z→w 8x512 FC embedding to the logical limit, and souping it up with self-attention instead of regular FC layers. It's one of the weirdest parts of StyleGAN, but seems important in generating a really useful linear disentangled w seed. The final output could also be done with a single pass into Reformer, which scales up to 256^2 = 65536 pixels and could potentially do 512^2 = 262144 pixels (this is easier than the usual sequence task because it only needs to 'decode' the latent state, like a bottleneck).

Another interesting option towards all-attention GANs: combine pixel-level loss, implemented using U-Net for pixel segmentation (["A U-Net Based Discriminator for Generative Adversarial Networks"](https://arxiv.org/abs/2002.12655), Schönfeld et al 2020) with FB's new simpler object detector, ["DETR: End-to-End Object Detection with Transformers"](https://arxiv.org/abs/2005.12872), Carion et al 2020.

Back in early March 2020, shawwn tested the simplest most obvious thing of just replacing (most of) the convolution layers in a 128px StyleGAN with self-attention, and it does still work with anime faces, so no immediate problem there...

**Comment by gwern** [comment]
There has been a rash of new papers from all the giants about various kinds of approximate or efficient attention mechanisms which may permit scaling to images. @lucidrains has abandoned Reformer/Sinkhorn/k-means which have proven frustrating to get working as advertised, and has been exploring the latest variants with [linear complexity recently proposed](https://github.com/lucidrains/linear-attention-transformer/) (they all make slightly different modifications to the softmax normalization around `QK` in self-attention), and thinks an [all-attention GAN may be doable right now](https://github.com/lucidrains/all-attention-gan) (As he has been using unconditional datasets like Flowers, I propose calling such an architecture a All-Attention Generative Unconditional Network, or AA-GUN.) 

Image sample from prototype (a small StyleGAN 2 run with linear attention on every generator and discriminator block):

![76-ema](https://user-images.githubusercontent.com/352559/84336872-3b696c00-ab66-11ea-9c4e-ce5673c83349.jpg)

**Comment by gwern** [comment]
One point about all-attention GANs I haven't mentioned: aside from the simplicity and potential power & scaling of attention layers, it ought to enable very easy multi-modal learning. Integrating text into BigGAN or StyleGAN is a real PITA right now. With attention layers, you just dump in text as part of the input sequence, no big deal. You have other metadata or stuff like depth maps? Dump it in. It's all sequences!

**Comment by hepesu** [comment]
Some results from my recent experiments on GANs based on TR. The TR is very smart and lazy, it may generate random samples neglecting the latent Z. For using latent after TR decoding, this model is similar to VQ-VAE, as that we can consider the latent code comes from some dicts like VQ-VAE. The following MLP transforms the latent code into pixel based on coordinates. But it was disappointing that the result was worse than VQ-VAE in my experiment.

**Comment by gwern** [comment]
["SMYRF: Efficient Attention using Asymmetric Clustering"](https://arxiv.org/abs/2010.05315), Daras et al 2020 experiments with attention more:

> We propose a novel type of balanced clustering algorithm to approximate attention. Attention complexity is reduced from O(N^2^) to O(N log N), where N is the sequence length. Our algorithm, SMYRF, uses Locality Sensitive Hashing (LSH) in a novel way by defining new Asymmetric transformations and an adaptive scheme that produces balanced clusters. The biggest advantage of SMYRF is that it can be used as a drop-in replacement for dense attention layers without any retraining. On the contrary, prior fast attention methods impose constraints (e.g. queries and keys share the same vector representations) and require re-training from scratch. We apply our method to pre-trained state-of-the-art Natural Language Processing and Computer Vision models and we report significant memory and speed benefits. Notably, SMYRF-BERT outperforms (slightly) BERT on GLUE, while using 50% less memory. We also show that SMYRF can be used interchangeably with dense attention before and after training. Finally, *we use SMYRF to train GANs with attention in high resolutions. Using a single TPU, we were able to scale attention to 128x128=16k and 256x256=65k tokens on BigGAN on CelebA-HQ.*
